cloud/scope/managedcontrolplane.go

```diff
@@ -0,0 +1,104 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scope
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/klogr"
+	azexpv1 "sigs.k8s.io/cluster-api-provider-azure/exp/api/v1alpha3"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	expv1 "sigs.k8s.io/cluster-api/exp/api/v1alpha3"
+
+	"sigs.k8s.io/cluster-api/util/patch"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// ManagedControlPlaneScopeParams defines the input parameters used to create a new Scope.
+type ManagedControlPlaneScopeParams struct {
+	AzureClients
+	Client           client.Client
+	Logger           logr.Logger
+	Cluster          *clusterv1.Cluster
+	ControlPlane     *azexpv1.AzureManagedControlPlane
+	Context          context.Context
+	InfraMachinePool *azexpv1.AzureManagedMachinePool
+	MachinePool      *expv1.MachinePool
+	PatchTarget      runtime.Object
+}
+
+// NewManagedControlPlaneScope creates a new Scope from the supplied parameters.
+// This is meant to be called for each reconcile iteration.
+func NewManagedControlPlaneScope(ctx context.Context, params ManagedControlPlaneScopeParams) (*ManagedControlPlaneScope, error) {
+	if params.Cluster == nil {
+		return nil, errors.New("failed to generate new scope from nil Cluster")
+	}
+
+	if params.ControlPlane == nil {
+		return nil, errors.New("failed to generate new scope from nil ControlPlane")
+	}
+
+	if params.Logger == nil {
+		params.Logger = klogr.New()
+	}
+
+	err := params.AzureClients.setCredentials()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create Azure session")
+	}
+
+	helper, err := patch.NewHelper(params.PatchTarget, params.Client)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to init patch helper")
+	}
+
+	return &ManagedControlPlaneScope{
+		Context:          ctx,
+		Logger:           params.Logger,
+		Client:           params.Client,
+		AzureClients:     params.AzureClients,
+		Cluster:          params.Cluster,
+		ControlPlane:     params.ControlPlane,
+		MachinePool:      params.MachinePool,
+		InfraMachinePool: params.InfraMachinePool,
+		PatchTarget:      params.PatchTarget,
+		patchHelper:      helper,
+	}, nil
+}
+
+// ManagedControlPlaneScope defines the basic context for an actuator to operate upon.
+type ManagedControlPlaneScope struct {
+	Context context.Context
+	logr.Logger
+	Client      client.Client
+	patchHelper *patch.Helper
+
+	AzureClients
+	Cluster          *clusterv1.Cluster
+	MachinePool      *expv1.MachinePool
+	ControlPlane     *azexpv1.AzureManagedControlPlane
+	InfraMachinePool *azexpv1.AzureManagedMachinePool
+	PatchTarget      runtime.Object
+}
+
+// PatchObject persists the cluster configuration and status.
+func (s *ManagedControlPlaneScope) PatchObject() error {
```

Comment:
This should take a context.

Reply:
https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/482#discussion_r420689947